src/librustc_typeck/check/mod.rs

```diff
@@ -1294,6 +1298,33 @@ pub fn check_simd<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, sp: Span, def_id: DefId
     }
 }
 
+fn has_align<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, root_def_id: DefId, def_id: DefId) -> bool {
```

Comment:
Wait, this only applies to nested `struct`s? That explains why no regressions, *however*, you probably want to handle it so you don't infinitely recurse in the compiler when some other part of the compiler wants to report a non-fatal error about it (see `AdtDef::calculate_sized_constraint_inner`, you probably don't need the cache though, just the stack to check for cycles).